File: nanak_customization/nanak_customization/page/stock_query/stock_query.js
frappe.pages['stock-query'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Stock Query',
		single_column: true
	});

	new erpnext.StockQuery(page);
}

erpnext.StockQuery = class StockQuery {

	constructor(page) {
		this.page = page;
		this.make_form();
	}

	make_form() {
		this.form = new frappe.ui.FieldGroup({
			fields: [
				{
					label: __('Item Code'),
					fieldname: 'item_code',
					fieldtype: 'Link',
					options: 'Item',
					change: async () => {
						this.fetch_and_render()
					},
					get_query: () => {
						return {
							filters: {
								"disabled": ["!=", 1]
							}
						}
					},
				},
				{
					label: __('UOM'),
					fieldname: 'stock_uom',
					fieldtype: 'Data',
					
					
					read_only:1
					
				},
				{
					fieldtype: 'Column Break'
				},
				{
					label: __('Item Name'),
					fieldname: 'item_name',
					fieldtype: 'Data',
					
					change: () => {},
					read_only:1
					
				},
				{
					label: __('Item Group'),
					fieldname: 'item_group',
					fieldtype: 'Link',
					options: 'Item Group',
					change: () => {},
					read_only:1
					
				},
				{
					fieldtype: 'Column Break'
				},
				{
					label: __('Stock Category'),
					fieldname: 'stock_category',
					fieldtype: 'Data',
					
					change: () => {},
					read_only:1
					
				},
				{
					label: __('Company Code'),
					fieldname: 'company_code',
					fieldtype: 'Data',
					
					change: () => {},
					read_only:1
					
				},
				{
					fieldtype: 'Column Break'
				},
				
				{
					label: __('Price'),
					fieldname: 'price',
					fieldtype: 'Currency',
					
					change: () => {},
					read_only:1
					
				},
				
				{
					label: __('Qty'),
					fieldname: 'qty',
					fieldtype: 'Float',
					
					change: () => {},
					read_only:1
					
				},
				{
					fieldtype: 'Column Break'
				},
				{
					label: __('Tax'),
					fieldname: 'tax',
					fieldtype: 'Data',
					
					change: () => {},
					read_only:1
					
				},
				{
					label: __('HSN'),
					fieldname: 'gst_hsn_code',
					fieldtype: 'Data',
					
					change: () => {},
					read_only:1
					
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Godownwise Stock",
					fieldtype: 'HTML',
					fieldname: 'stock'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Similar Items",
					fieldtype: 'HTML',
					fieldname: 'similar_items'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Sales",
					fieldtype: 'HTML',
					fieldname: 'sales'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Purchase",
					fieldtype: 'HTML',
					fieldname: 'purchase'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Pending PO",
					fieldtype: 'HTML',
					fieldname: 'pending_po'
				},
				{
					fieldtype: 'Section Break'
				},
				{
					label:"Pending SO",
					fieldtype: 'HTML',
					fieldname: 'pending_so'
				}
			],
			body: this.page.body
		});
		this.form.make();
		$(this.form.wrapper)
				.find(".input-with-feedback")
				.css({
					backgroundColor: '#cfcfcf'
				});
	}
	fetch_and_render(){
		// console.log(this.form.get_value("item_code"));
		var item_code = this.form.get_value("item_code");
		this.set_header_values(item_code)
		this.set_past_purchase(item_code)
		this.set_past_sales(item_code)
		this.set_similar_stock(item_code)
		this.set_warehouse_stock(item_code)
		this.set_pending_po(item_code)
		this.set_pending_so(item_code)
	}
	async set_header_values(item_code){
		// console.log(this.form.fields[1])
		var res = await get_header_list(item_code)
		// console.log(res)
		this.form.set_value("stock_uom",res.message[0].stock_uom)
		this.form.set_value("stock_category",res.message[0].stock_category)
		this.form.set_value("company_code",res.message[0].company_code)
		this.form.set_value("item_group",res.message[0].item_group)
		this.form.set_value("item_name",res.message[0].item_name)
		this.form.set_value("price",res.message[0].price)
		this.form.set_value("tax",res.message[0].item_tax_template)
		this.form.set_value("gst_hsn_code",res.message[0].gst_hsn_code)
		this.form.set_value("qty",res.message[0].qty)
	}
	async set_warehouse_stock(item_code){
		if(item_code){

		var res = await get_warehouse_stock(item_code)
		console.log(res)
		var html_content = '';
		html_content += '<div class="frappe-card"><h4>Warehouse Wise Stock</h4><table class="table"><thead><tr><th>Warehouse</th><th>Batch/Serial No</th><th>Open Qty</th><th>Reserve Qty</th></tr></thead><tbody>';
		res.message[0].forEach(i =>{
			console.log(i)
			if(res.message[1].has_serial_no){
				html_content +=
			'<tr>' +
				'<td>' + i.warehouse +'</td>' +
				'<td>' + i.serial_no +' </td>' +
				'<td>' + i.qty +'</td>' +
				'<td>' + i.reserved +' </td>' +
				
			'</tr>';
			}else if(res.message[1].has_batch_no){
				html_content +=
			'<tr>' +
				'<td>' + i.warehouse +'</td>' +
				'<td>' + i.batch_id +' </td>' +
				'<td>' + i.qty +'</td>' +
				'<td>' + i.reserved +' </td>' +
				
			'</tr>';
			}else{
				html_content +=
			'<tr>' +
				'<td>' + i.warehouse +'</td>' +
				'<td></td>' +
				'<td>' + i.qty +'</td>' +
				'<td>' + i.reserved +' </td>' +
				
			'</tr>';
			}
		})
		

		html_content += '</tbody></table></div>'

		this.form.get_field('stock').html(html_content);

		}
		
	}
	async set_similar_stock(item_code){
		var res = await get_similar_items(item_code)
		// console.log(res.message)
		var html_content = '';
		html_content += '<div class="frappe-card">' +
			'<h4>Similar Items</h4>' +
			'<table class="table">' +
			'<thead>' +
			'<tr>' +
				'<th>Item Code</th>' +
				'<th>Item Name</th>' +
				'<th>Qty</th>' +
			'</tr>' +
			'</thead>' +
			'<tbody>';
		for(var i in res.message){
			var link = 'item/' + encodeURI(res.message[i].name)
			html_content +=
			'<tr>' +
				'<td><a onclick="frappe.utils.copy_to_clipboard(\'' + res.message[i].name + '\')">' + res.message[i].name + '</a></td>' +
				'<td>' + res.message[i].item_name +'</td>' +
				'<td>' + res.message[i].qty +'</td>' +
			'</tr>';
		}
		
		html_content +=
			'</tbody>' +
			'</table>' +
			'</div>';

		this.form.get_field('similar_items').html(html_content);
	}
	async set_past_sales(item_code){
		var res = await get_sales_order(item_code)
		var html_content = '';

		html_content += '<div class="frappe-card"><h4>Past Sales</h4><table class="table"><thead><tr><th>Bill No</th><th>Bill Date</th><th>Party Name</th><th>Unit</th><th>Qty</th><th>Item Value</th></tr></thead><tbody>';

		for(var i in res.message){
			var link = 'sales-invoice/' + encodeURI(res.message[i].parent)
			html_content +=
			'<tr>' +
				'<td><a href=' + link.toString() + '>' + res.message[i].parent + '</a></td>' +
				'<td>' + res.message[i].creation +'</td>' +
				'<td>' + res.message[i].customer +'</td>' +
				'<td>' + res.message[i].uom +'</td>' +
				'<td>' + res.message[i].qty +'</td>' +
				'<td>' + res.message[i].rate +'</td>' +
			'</tr>';
		}

		html_content += '</tbody></table></table></div>';

		this.form.get_field('sales').html(html_content);
	}
	async set_past_purchase(item_code){
		var res = await get_purchase_order(item_code)

		var html_content = '';

		html_content += '<div class="frappe-card"><h4>Past Purchase</h4><table class="table"><thead><tr><th>Bill No</th><th>Bill Date</th><th>Party Name</th><th>Unit</th><th>Qty</th><th>Item Value</th></tr></thead><tbody>';

		for(var i in res.message){
			var link = 'purchase-invoice/' + encodeURI(res.message[i].parent)
			html_content +=
			'<tr>' +
				'<td><a href=' + link.toString() + '>' + res.message[i].parent + '</a></td>' +
				'<td>' + res.message[i].creation +'</td>' +
				'<td>' + res.message[i].supplier +'</td>' +
				'<td>' + res.message[i].uom +'</td>' +
				'<td>' + res.message[i].qty +'</td>' +
				'<td>' + res.message[i].rate +'</td>' +
			'</tr>';
		}

		html_content += '</tbody></table></table></div>';

		this.form.get_field('purchase').html(html_content);
	}
	async set_pending_so(item_code){
		var res = await get_pending_so(item_code)

		var html_content = '';

		html_content += '<div class="frappe-card"><h4>Pending Sales</h4><table class="table"><thead><tr><th>Bill No</th><th>Bill Date</th><th>Party Name</th><th>Unit</th><th>Qty</th><th>Item Value</th></tr></thead><tbody>';

		for(var i in res.message){
			var link = 'sales-order/' + encodeURI(res.message[i].parent)
			html_content +=
			'<tr>' +
				'<td><a href=' + link.toString() + '>' + res.message[i].parent + '</a></td>' +
				'<td>' + res.message[i].creation +'</td>' +
				'<td>' + res.message[i].customer +'</td>' +
				'<td>' + res.message[i].uom +'</td>' +
				'<td>' + res.message[i].qty +'</td>' +
				'<td>' + res.message[i].rate +'</td>' +
			'</tr>';
		}

		html_content += '</tbody></table></table></div>';

		this.form.get_field('pending_so').html(html_content);
	}
	async set_pending_po(item_code){
		var res = await get_pending_po(item_code)

		var html_content = '';

		html_content += '<div class="frappe-card"><h4>Pending Purchase</h4><table class="table"><thead><tr><th>Bill No</th><th>Bill Date</th><th>Party Name</th><th>Unit</th><th>Qty</th><th>Item Value</th></tr></thead><tbody>';

		for(var i in res.message){
			var link = 'purchase-order/' + encodeURI(res.message[i].parent)
			html_content +=
			'<tr>' +
				'<td><a href=' + link.toString() + '>' + res.message[i].parent + '</a></td>' +
				'<td>' + res.message[i].creation +'</td>' +
				'<td>' + res.message[i].customer +'</td>' +
				'<td>' + res.message[i].uom +'</td>' +
				'<td>' + res.message[i].qty +'</td>' +
				'<td>' + res.message[i].rate +'</td>' +
			'</tr>';
		}

		html_content += '</tbody></table></table></div>';

		this.form.get_field('pending_po').html(html_content);
	}
}


function get_header_list(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_header_data',
				'args': {
					'item_code': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_similar_items(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_same_category',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_sales_order(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_sales_order',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_purchase_order(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_purchase_order',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_pending_po(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_pending_po',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_pending_so(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_pending_so',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

function get_warehouse_stock(item) {
	return new Promise(function(resolve, reject){
		try{
			frappe.call({
				'method': 'nanak_customization.nanak_customization.page.stock_query.stock_query.get_godown_wise',
				'args': {
					'product_id': item
				},
				callback: resolve
			});
		} catch (e) {reject(e);}
	});
}

// function get_list (opts) {
// 	return new Promise(function (resolve, reject) {
// 		try {
// 			frappe.call({
// 				method: FRAPPE_CLIENT + '.get_list',
// 				args: {
// 					doctype: opts.doctype,
// 					fields: opts.fields,
// 					filters: opts.filters,
// 					order_by: opts.order_by,
// 					limit_start: opts.limit_start,
// 					limit_page_length: opts.limit_page_length,
// 					parent: opts.parent
// 				},
// 				callback: resolve
// 			});
// 		} catch (e) { reject(e); }
// 	});
// }


// refresh : async function(frm){
// let list_data = await get_list(ops)

// }

File: nanak_customization/nanak_customization/page/stock_query/stock_query.py
# Copyright (c) 2022, Akhilam and contributors
# For license information, please see license.txt

import frappe
from frappe.model.document import Document
from erpnext.stock.doctype.batch.batch import get_batch_no,get_batches
# from console import console

class QueryTesting(Document):
	pass
	

@frappe.whitelist()
def get_header_data(item_code):
		# item_code = 'Glass'

		pricelist = frappe.db.get_single_value("Nanak Standard Settings", "stock_query_pricelist")

		data = frappe.db.sql("""
			select
			i.item_name,
			i.item_group,i.stock_category,i.company_code,it.item_tax_template,i.stock_uom,
			(select ip.price_list_rate from `tabItem Price` ip where ip.item_code = i.name and price_list = %s order by creation desc limit 1) as price,
			i.gst_hsn_code,
			(select sum(sle.actual_qty) from `tabBin` sle where sle.item_code = i.name and sle.warehouse in (select W.name from `tabWarehouse` W where W.is_reserve_warehouse = 0) group by sle.item_code) as qty
			from
			`tabItem` i left join `tabItem Tax` it on i.name = it.parent
			where
			i.name = %s
			order by it.valid_from desc
		""", (pricelist,item_code), as_dict = True)

		return data

@frappe.whitelist()
def get_same_category(product_id):
		# product_id = 'Mouse'
		# if want to omit product mouse
		ext = " and i.name != '" + product_id + "'"

		product_group = frappe.db.get_value('Item', product_id, 'stock_category')

		data = frappe.db.sql("""
			select
			i.name,
			i.item_name,
			i.item_group,
			(select sum(sle.actual_qty) from `tabBin` sle where sle.item_code = i.name and sle.warehouse in (select W.name from `tabWarehouse` W where W.is_reserve_warehouse = 0) group by sle.item_code) as qty
			from
			`tabItem` i
			where
			i.stock_category = '%s'
			%s
			
		"""% (product_group, ext), as_dict = True)

		return data

@frappe.whitelist()
def get_godown_wise(product_id):
	# frappe.msgprint("fetch warehouse data")
	warehouse_data = []
	sn_bn = frappe.db.get_value("Item",product_id,["has_serial_no","has_batch_no"],as_dict=1)

	if sn_bn['has_serial_no']:
		actual_data = frappe.db.sql("""
		select wh.is_reserve_warehouse,wh.warehouse as og_warehouse,sn.serial_no,wh.name as warehouse from  `tabWarehouse` wh inner join `tabSerial No` sn on wh.name = sn.warehouse where sn.item_code = %s and sn.status = 'Active'
		""",product_id,as_dict = 1)
		for item in actual_data:
			if item.is_reserve_warehouse:
				warehouse_data.append({
					"warehouse":item.og_warehouse,
					"qty":0,
					"serial_no":item.serial_no,
					"reserved":1
				})
			else:
				warehouse_data.append({
					"warehouse":item.warehouse,
					"qty":1,
					"serial_no":item.serial_no,
					"reserved":0
				})
		return warehouse_data,sn_bn
	if sn_bn['has_batch_no']:
		actual_data = frappe.db.sql("""
		select bn.warehouse,bn.actual_qty,bn.item_code,wh.is_reserve_warehouse,wh.warehouse as og_warehouse from `tabBin` bn inner join `tabWarehouse` wh on wh.name = bn.warehouse where bn.actual_qty > 0 and bn.item_code = %s
		""",product_id,as_dict = 1)
		# console(actual_data).log()
		raw_batch_data = []
		warehouse_data = []
		for item in actual_data:
			batch = get_batches(item.item_code, item.warehouse)
			for b in batch:
				b['warehouse'] = item.warehouse	
				b['is_reserve_warehouse'] = item.is_reserve_warehouse
				b['og_warehouse'] = item.og_warehouse
				raw_batch_data.append(b)
		for item in raw_batch_data:
			if item.is_reserve_warehouse:
				# console(item).log()
				
				
				if any(d['warehouse'] == item.og_warehouse and d['batch_id'] == item.batch_id for d in warehouse_data):
					# console("modify reserve").log()
					
					for data in warehouse_data:
						if data['warehouse'] == item.og_warehouse and data['batch_id'] == item.batch_id:
							data['reserved'] = item.qty
				else:
					# console("append reserve").log()
					warehouse_data.append({
					"warehouse":item.og_warehouse,
					"qty":0,
					"reserved":item.qty,
					"batch_id":item.batch_id
				})
			else:
				if any(d['warehouse'] == item.warehouse and d['batch_id'] == item.batch_id for d in warehouse_data):
					# console("modify qty").log()
					# console(item).log()
					for data in warehouse_data:
						if data['warehouse'] == item.warehouse and data['batch_id'] == item.batch_id:
							data['qty'] = item.qty
				else:
					# console("append qty").log()
					warehouse_data.append({
						"warehouse":item.warehouse,
						"qty":item.qty,
						"batch_id":item.batch_id,
						"reserved":0
					})
			# console(warehouse_data).log()


		# frappe.throw(str(raw_batch_data))
		return warehouse_data,sn_bn

	else:
		actual_data = frappe.db.sql("""
		select bn.warehouse,bn.actual_qty,wh.is_reserve_warehouse,wh.warehouse as og_warehouse from `tabBin` bn inner join `tabWarehouse` wh on wh.name = bn.warehouse where bn.actual_qty > 0 and bn.item_code = %s 
		""",product_id,as_dict = 1)
		
		for item in actual_data:
			if item.is_reserve_warehouse:
				if any(d['warehouse'] == item.og_warehouse for d in warehouse_data):
					for data in warehouse_data:
						if data['warehouse'] == item.og_warehouse:
							data['reserved'] = item.actual_qty

				else:
					warehouse_data.append({
					"warehouse":item.og_warehouse,
					"qty":0,
					"reserved":item.actual_qty
				})
			else:
				warehouse_data.append({
					"warehouse":item.warehouse,
					"qty":item.actual_qty,
					"reserved":0
				})
			
		return warehouse_data,sn_bn




	# 
	

	# warehouse_data = []
	# for item in actual_data:
	# 	if item.is_reserve_warehouse:
	# 		if any(d['warehouse'] == item.og_warehouse for d in warehouse_data):
	# 			for data in warehouse_data:
	# 				if data['warehouse'] == item.og_warehouse:
	# 					data['reserved'] = item.actual_qty

	# 		else:
	# 			warehouse_data.append({
	# 			"warehouse":item.og_warehouse,
	# 			"qty":0,
	# 			"reserved":item.actual_qty,
	# 			"serial_no":"",
	# 			"batch_no":"",
	# 			"has_serial_no":item.has_serial_no,
	# 			"has_batch_no":item.has_batch_no
	# 		})
				
	# 	else:	
	# 		serial_no_str=""
	# 		batch_no_str=""
	# 		# frappe.msgprint(str(item.has_serial_no))
	# 		# frappe.msgprint(str(item.has_batch_no))
	# 		if item.has_serial_no:	
	# 			frappe.msgprint("serial no")		
	# 			serial_no = frappe.db.sql("""
	# 			select name from `tabSerial No` where status = 'Active' and item_code = %s and warehouse = %s
	# 			""",(product_id,item.warehouse),as_dict = 1)
	# 			# frappe.msgprint(str(serial_no))
	# 			serial_no_str = ",".join([item['name'] for item in serial_no])
	# 			# frappe.msgprint(serial_no_str)
	# 		if item.has_batch_no:
	# 			# frappe.msgprint("batch no")
	# 			batch = get_batches(product_id, item.warehouse)

	# 			batch_no_str = ",".join([str(item['batch_id'])+" | <b>"+str(item['qty'])+"</b>" for item in batch])
	# 			# frappe.throw(str(batch))

	# 		warehouse_data.append({
	# 			"warehouse":item.warehouse,
	# 			"qty":item.actual_qty,
	# 			"reserved":0,
	# 			"serial_no":serial_no_str,
	# 			"batch_no":batch_no_str,
	# 			"has_serial_no":item.has_serial_no,
	# 			"has_batch_no":item.has_batch_no

	# 		})
	# return warehouse_data

def prepare_warehousewise_data(sn_bn,data):
	warehouse_data = []
	if sn_bn[0]:
		pass
	elif sn_bn[1]:
		for item in data:
			if item.is_reserve_warehouse:
				if any(d['warehouse'] == item.og_warehouse for d in warehouse_data):
					for data in warehouse_data:
						if data['warehouse'] == item.og_warehouse:
							data['reserved'] = item.actual_qty

				else:
					warehouse_data.append({
					"warehouse":item.og_warehouse,
					"qty":0,
					"reserved":item.actual_qty
				})
			else:
				warehouse_data.append({
					"warehouse":item.warehouse,
					"qty":item.actual_qty,
					"reserved":0
				})
		

	else:
		for item in data:
			if item.is_reserve_warehouse:
				if any(d['warehouse'] == item.og_warehouse for d in warehouse_data):
					for data in warehouse_data:
						if data['warehouse'] == item.og_warehouse:
							data['reserved'] = item.actual_qty

				else:
					warehouse_data.append({
					"warehouse":item.og_warehouse,
					"qty":0,
					"reserved":item.actual_qty
				})
			else:
				warehouse_data.append({
					"warehouse":item.warehouse,
					"qty":item.actual_qty,
					"reserved":0
				})
		
		
		



		
		
	# frappe.throw(warehouse_data)


	# warehouse = frappe.db.get_single_value("Stock Settings", "default_warehouse")

	# data = frappe.db.sql("""
	# 	select 
	# 	i.name,
	# 	i.item_name,
	# 	sle.qty_after_transaction,
	# 	sle.warehouse
	# 	from
	# 	`tabItem` i
	# 	left join (
	# 	select item_code, qty_after_transaction, warehouse
	# 	from
	# 	`tabStock Ledger Entry`
	# 	group by item_code, warehouse
	# 	order by creation desc
	# 	) sle on sle.item_code = i.name
	# 	where
	# 	sle.warehouse <> %s
	# 	and
	# 	i.name = %s
		
	# """, (warehouse, product_id), as_dict = True)

	# return data

@frappe.whitelist()
def get_sales_order(product_id):
	# product_id = 'Tubelight'

	data = frappe.db.sql("""
		select
		sii.creation,
		sii.parent,
		sii.uom,
		sii.qty,
		sii.rate,
		si.customer
		from
		`tabSales Invoice Item` sii
		left join `tabSales Invoice` si on si.name = sii.parent
		where item_code = %s
		and si.docstatus = 1
		order by creation desc
		limit 6
		
	""", (product_id), as_dict = True)
	
	return data

@frappe.whitelist()
def get_purchase_order(product_id):
	# product_id = 'Phone'

	data = frappe.db.sql("""
		select
		pii.creation,
		pii.parent,
		pii.uom,
		pii.qty,
		pii.rate,
		pi.supplier
		from
		`tabPurchase Invoice Item` pii
		left join `tabPurchase Invoice` pi on pi.name = pii.parent
		where pii.item_code = %s
		and pi.docstatus = 1
		order by creation desc
		limit 6

		""", (product_id),as_dict = True)
	return data

@frappe.whitelist()
def get_pending_po(product_id):
	# product_id = 'Phone'

	data = frappe.db.sql("""
		select
		poi.creation,
		poi.parent,
		poi.uom,
		poi.qty,
		poi.rate,
		po.supplier
		from
		`tabPurchase Order Item` poi
		left join `tabPurchase Order` po on po.name = poi.parent
		where poi.item_code = %s
		and po.docstatus = 1
		and po.name not in (select purchase_order from `tabPurchase Invoice Item`)
		order by creation desc
		limit 6

		""", (product_id),as_dict = True)
	return data

@frappe.whitelist()
def get_pending_so(product_id):
	# product_id = 'Phone'

	data = frappe.db.sql("""
		select
		soi.creation,
		soi.parent,
		soi.uom,
		soi.qty,
		soi.rate,
		so.customer
		from
		`tabSales Order Item` soi
		left join `tabSales Order` so on so.name = soi.parent
		where soi.item_code = %s
		and so.docstatus = 1
		and so.name not in (select sales_order from `tabSales Invoice Item`)
		order by creation desc
		limit 6
		
	""", (product_id),as_dict = True)
	return data
